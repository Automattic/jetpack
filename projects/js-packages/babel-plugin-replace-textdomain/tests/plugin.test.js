/* eslint-disable import/order */

const mockOrigDebug = jest.requireActual( 'debug' );
const mockDebug = jest.fn();
jest.mock( 'debug', () => {
	return name => {
		if ( name.startsWith( '@automattic/babel-plugin-replace-textdomain' ) ) {
			return mockDebug;
		}
		return mockOrigDebug( name );
	};
} );
const setup = () => {
	mockDebug.mockClear();
	return () => {
		expect( mockDebug.mock.calls ).toMatchSnapshot( 'debug calls' );
	};
};

const pluginTester = require( 'babel-plugin-tester' ).default;
const plugin = require( '../src/index.js' );

pluginTester( {
	plugin,
	pluginName: require( '../package.json' ).name,
	filename: __filename,
	snapshot: true,
	babelOptions: {
		highlightCode: false,
	},
	tests: [
		{
			title: 'Simple test',
			setup,
			fixture: 'fixtures/simple.js',
			pluginOptions: {
				textdomain: 'new-domain',
			},
		},
		{
			title: 'Multiple domains, constant mapping',
			setup,
			fixture: 'fixtures/multi.js',
			pluginOptions: {
				textdomain: 'new-domain',
			},
		},
		{
			title: 'Multiple domains, object mapping',
			setup,
			fixture: 'fixtures/multi.js',
			pluginOptions: {
				textdomain: {
					a: 'AAA',
					c: 'CCC',
				},
			},
		},
		{
			title: 'Multiple domains, function mapping',
			setup,
			fixture: 'fixtures/multi.js',
			pluginOptions: {
				textdomain: d => d.toUpperCase(),
			},
		},
		{
			title: 'Custom functions list',
			setup,
			fixture: 'fixtures/simple.js',
			pluginOptions: {
				textdomain: 'new-domain',
				functions: {
					_not: 3,
				},
			},
		},

		{
			title: 'Missing domain parameter',
			setup,
			code: `__( 'No domain' );`,
			pluginOptions: {
				textdomain: 'new-domain',
			},
		},
		{
			title: 'Missing domain parameter, no replacement',
			setup,
			code: `__( 'No domain' );`,
			pluginOptions: {
				textdomain: () => null,
			},
		},
		{
			title: 'Missing context and domain parameters',
			setup,
			code: `_x( 'No domain' );`,
			pluginOptions: {
				textdomain: 'new-domain',
			},
		},
		{
			title: 'Missing context and domain parameters, no replacement',
			setup,
			code: `_x( 'No domain' );`,
			pluginOptions: {
				textdomain: () => null,
			},
		},
		{
			title: 'Non-literal domain',
			setup,
			code: `__( 'Non-literal domain', domain );`,
			pluginOptions: {
				textdomain: 'new-domain',
			},
		},
		{
			title: 'Template-string domain',
			setup,
			code: "__( 'Template-string domain', `domain` );",
			pluginOptions: {
				textdomain: 'new-domain',
			},
		},
		{
			title: 'Template-string domain, no replacement',
			setup,
			code: "__( 'Template-string domain', `domain` );",
			pluginOptions: {
				textdomain: () => null,
			},
		},
		{
			title: 'Template-string domain with expression',
			setup,
			code: "__( 'Template-string domain', `domain ${ x }` );",
			pluginOptions: {
				textdomain: 'new-domain',
			},
		},
		{
			title: 'Expression domain',
			setup,
			code: `__( 'Expression', 'dom' + 'ain' );`,
			pluginOptions: {
				textdomain: 'new-domain',
			},
		},

		{
			title: "Doesn't try to handle `toString()` or the like",
			setup,
			code: `x.toString();`,
			snapshot: false,
			pluginOptions: {
				textdomain: 'new-domain',
			},
		},

		// Invalid option handling.
		{
			title: 'Bad options: missing textdomain',
			fixture: 'fixtures/simple.js',
			pluginOptions: {},
			snapshot: false,
			error: 'The `textdomain` option is not set.',
		},
		{
			title: 'Bad options: bad textdomain',
			fixture: 'fixtures/simple.js',
			pluginOptions: {
				textdomain: 123,
			},
			snapshot: false,
			error: 'The `textdomain` option is set to an invalid value.',
		},
		{
			title: 'Bad options: bad functions',
			fixture: 'fixtures/simple.js',
			pluginOptions: {
				textdomain: 'foo',
				functions: [],
			},
			snapshot: false,
			error: 'The `functions` option is set to an invalid value.',
		},
		{
			title: 'Bad options: bad value in functions',
			fixture: 'fixtures/simple.js',
			pluginOptions: {
				textdomain: 'foo',
				functions: {
					__: 1,
					_x: '2',
				},
			},
			snapshot: false,
			error: 'Invalid argument index for `functions._x`, value must be a non-negative integer.',
		},
		{
			title: 'Bad options: negative value in functions',
			fixture: 'fixtures/simple.js',
			pluginOptions: {
				textdomain: 'foo',
				functions: { __: -1 },
			},
			snapshot: false,
			error: 'Invalid argument index for `functions.__`, value must be a non-negative integer.',
		},
		{
			title: 'Bad options: non-integer value in functions',
			fixture: 'fixtures/simple.js',
			pluginOptions: {
				textdomain: 'foo',
				functions: { __: 1.0001 },
			},
			snapshot: false,
			error: 'Invalid argument index for `functions.__`, value must be a non-negative integer.',
		},
	],
} );

test( 'Default functions exported', () => {
	expect( plugin.defaultFunctions ).toBeDefined();

	// Test that it's immutable.
	plugin.defaultFunctions.__ = 10;
	plugin.defaultFunctions.foo = 11;
	expect( plugin.defaultFunctions.__ ).toBe( 1 );
	expect( plugin.defaultFunctions.foo ).toBeUndefined();
} );
