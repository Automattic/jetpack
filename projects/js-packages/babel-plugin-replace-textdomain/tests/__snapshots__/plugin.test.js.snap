// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@automattic/babel-plugin-replace-textdomain 1. Simple test: 1. Simple test 1`] = `

__( 'foo', 'srcdomain' );
_x( 'foo', 'context', 'srcdomain' );
_n( 'foo', 'foos', n, 'srcdomain' );
_nx( 'foo', 'foos', n, 'context', 'srcdomain' );

i18n.__( 'foo', 'srcdomain' );
i18n._x( 'foo', 'context', 'srcdomain' );
i18n._n( 'foo', 'foos', n, 'srcdomain' );
i18n._nx( 'foo', 'foos', n, 'context', 'srcdomain' );

_not( 'foo', 'foos', n, 'srcdomain' );
i18n._not( 'foo', 'foos', n, 'srcdomain' );
__.not( 'foo', 'srcdomain' );


      ↓ ↓ ↓ ↓ ↓ ↓

__('foo', 'new-domain');
_x('foo', 'context', 'new-domain');
_n('foo', 'foos', n, 'new-domain');
_nx('foo', 'foos', n, 'context', 'new-domain');
i18n.__('foo', 'new-domain');
i18n._x('foo', 'context', 'new-domain');
i18n._n('foo', 'foos', n, 'new-domain');
i18n._nx('foo', 'foos', n, 'context', 'new-domain');
_not('foo', 'foos', n, 'srcdomain');
i18n._not('foo', 'foos', n, 'srcdomain');
__.not('foo', 'srcdomain');

`;

exports[`@automattic/babel-plugin-replace-textdomain 1. Simple test: debug calls 1`] = `[]`;

exports[`@automattic/babel-plugin-replace-textdomain 2. Multiple domains, constant mapping: 2. Multiple domains, constant mapping 1`] = `

__( 'foo', 'a' );
_x( 'foo', 'context', 'b' );
_n( 'foo', 'foos', n, 'c' );
_nx( 'foo', 'foos', n, 'context', 'd' );

i18n.__( 'foo', 'a' );
i18n._x( 'foo', 'context', 'b' );
i18n._n( 'foo', 'foos', n, 'c' );
i18n._nx( 'foo', 'foos', n, 'context', 'd' );

_not( 'foo', 'foos', n, 'a' );
i18n._not( 'foo', 'foos', n, 'b' );
__.not( 'foo', 'c' );


      ↓ ↓ ↓ ↓ ↓ ↓

__('foo', 'new-domain');
_x('foo', 'context', 'new-domain');
_n('foo', 'foos', n, 'new-domain');
_nx('foo', 'foos', n, 'context', 'new-domain');
i18n.__('foo', 'new-domain');
i18n._x('foo', 'context', 'new-domain');
i18n._n('foo', 'foos', n, 'new-domain');
i18n._nx('foo', 'foos', n, 'context', 'new-domain');
_not('foo', 'foos', n, 'a');
i18n._not('foo', 'foos', n, 'b');
__.not('foo', 'c');

`;

exports[`@automattic/babel-plugin-replace-textdomain 2. Multiple domains, constant mapping: debug calls 1`] = `[]`;

exports[`@automattic/babel-plugin-replace-textdomain 3. Multiple domains, object mapping: 3. Multiple domains, object mapping 1`] = `

__( 'foo', 'a' );
_x( 'foo', 'context', 'b' );
_n( 'foo', 'foos', n, 'c' );
_nx( 'foo', 'foos', n, 'context', 'd' );

i18n.__( 'foo', 'a' );
i18n._x( 'foo', 'context', 'b' );
i18n._n( 'foo', 'foos', n, 'c' );
i18n._nx( 'foo', 'foos', n, 'context', 'd' );

_not( 'foo', 'foos', n, 'a' );
i18n._not( 'foo', 'foos', n, 'b' );
__.not( 'foo', 'c' );


      ↓ ↓ ↓ ↓ ↓ ↓

__('foo', 'AAA');
_x('foo', 'context', 'b');
_n('foo', 'foos', n, 'CCC');
_nx('foo', 'foos', n, 'context', 'd');
i18n.__('foo', 'AAA');
i18n._x('foo', 'context', 'b');
i18n._n('foo', 'foos', n, 'CCC');
i18n._nx('foo', 'foos', n, 'context', 'd');
_not('foo', 'foos', n, 'a');
i18n._not('foo', 'foos', n, 'b');
__.not('foo', 'c');

`;

exports[`@automattic/babel-plugin-replace-textdomain 3. Multiple domains, object mapping: debug calls 1`] = `
[
  [
    No mapping for textdomain b (first instance)
  1 | __( 'foo', 'a' );
> 2 | _x( 'foo', 'context', 'b' );
    |                       ^^^
  3 | _n( 'foo', 'foos', n, 'c' );
  4 | _nx( 'foo', 'foos', n, 'context', 'd' );
  5 |,
  ],
  [
    No mapping for textdomain d (first instance)
  2 | _x( 'foo', 'context', 'b' );
  3 | _n( 'foo', 'foos', n, 'c' );
> 4 | _nx( 'foo', 'foos', n, 'context', 'd' );
    |                                   ^^^
  5 |
  6 | i18n.__( 'foo', 'a' );
  7 | i18n._x( 'foo', 'context', 'b' );,
  ],
]
`;

exports[`@automattic/babel-plugin-replace-textdomain 4. Multiple domains, function mapping: 4. Multiple domains, function mapping 1`] = `

__( 'foo', 'a' );
_x( 'foo', 'context', 'b' );
_n( 'foo', 'foos', n, 'c' );
_nx( 'foo', 'foos', n, 'context', 'd' );

i18n.__( 'foo', 'a' );
i18n._x( 'foo', 'context', 'b' );
i18n._n( 'foo', 'foos', n, 'c' );
i18n._nx( 'foo', 'foos', n, 'context', 'd' );

_not( 'foo', 'foos', n, 'a' );
i18n._not( 'foo', 'foos', n, 'b' );
__.not( 'foo', 'c' );


      ↓ ↓ ↓ ↓ ↓ ↓

__('foo', 'A');
_x('foo', 'context', 'B');
_n('foo', 'foos', n, 'C');
_nx('foo', 'foos', n, 'context', 'D');
i18n.__('foo', 'A');
i18n._x('foo', 'context', 'B');
i18n._n('foo', 'foos', n, 'C');
i18n._nx('foo', 'foos', n, 'context', 'D');
_not('foo', 'foos', n, 'a');
i18n._not('foo', 'foos', n, 'b');
__.not('foo', 'c');

`;

exports[`@automattic/babel-plugin-replace-textdomain 4. Multiple domains, function mapping: debug calls 1`] = `[]`;

exports[`@automattic/babel-plugin-replace-textdomain 5. Custom functions list: 5. Custom functions list 1`] = `

__( 'foo', 'srcdomain' );
_x( 'foo', 'context', 'srcdomain' );
_n( 'foo', 'foos', n, 'srcdomain' );
_nx( 'foo', 'foos', n, 'context', 'srcdomain' );

i18n.__( 'foo', 'srcdomain' );
i18n._x( 'foo', 'context', 'srcdomain' );
i18n._n( 'foo', 'foos', n, 'srcdomain' );
i18n._nx( 'foo', 'foos', n, 'context', 'srcdomain' );

_not( 'foo', 'foos', n, 'srcdomain' );
i18n._not( 'foo', 'foos', n, 'srcdomain' );
__.not( 'foo', 'srcdomain' );


      ↓ ↓ ↓ ↓ ↓ ↓

__('foo', 'srcdomain');
_x('foo', 'context', 'srcdomain');
_n('foo', 'foos', n, 'srcdomain');
_nx('foo', 'foos', n, 'context', 'srcdomain');
i18n.__('foo', 'srcdomain');
i18n._x('foo', 'context', 'srcdomain');
i18n._n('foo', 'foos', n, 'srcdomain');
i18n._nx('foo', 'foos', n, 'context', 'srcdomain');
_not('foo', 'foos', n, 'new-domain');
i18n._not('foo', 'foos', n, 'new-domain');
__.not('foo', 'srcdomain');

`;

exports[`@automattic/babel-plugin-replace-textdomain 5. Custom functions list: debug calls 1`] = `[]`;

exports[`@automattic/babel-plugin-replace-textdomain 6. Missing domain parameter: 6. Missing domain parameter 1`] = `

__( 'No domain' );

      ↓ ↓ ↓ ↓ ↓ ↓

__('No domain', 'new-domain');

`;

exports[`@automattic/babel-plugin-replace-textdomain 6. Missing domain parameter: debug calls 1`] = `
[
  [
    Domain argument (index 2) is missing
> 1 | __( 'No domain' );
    | ^^^^^^^^^^^^^^^^^,
  ],
]
`;

exports[`@automattic/babel-plugin-replace-textdomain 7. Missing domain parameter, no replacement: 7. Missing domain parameter, no replacement 1`] = `

__( 'No domain' );

      ↓ ↓ ↓ ↓ ↓ ↓

__('No domain');

`;

exports[`@automattic/babel-plugin-replace-textdomain 7. Missing domain parameter, no replacement: debug calls 1`] = `
[
  [
    Domain argument (index 2) is missing
> 1 | __( 'No domain' );
    | ^^^^^^^^^^^^^^^^^,
  ],
]
`;

exports[`@automattic/babel-plugin-replace-textdomain 8. Missing context and domain parameters: 8. Missing context and domain parameters 1`] = `

_x( 'No domain' );

      ↓ ↓ ↓ ↓ ↓ ↓

_x('No domain', undefined, 'new-domain');

`;

exports[`@automattic/babel-plugin-replace-textdomain 8. Missing context and domain parameters: debug calls 1`] = `
[
  [
    Domain argument (index 3) is missing
> 1 | _x( 'No domain' );
    | ^^^^^^^^^^^^^^^^^,
  ],
]
`;

exports[`@automattic/babel-plugin-replace-textdomain 9. Missing context and domain parameters, no replacement: 9. Missing context and domain parameters, no replacement 1`] = `

_x( 'No domain' );

      ↓ ↓ ↓ ↓ ↓ ↓

_x('No domain');

`;

exports[`@automattic/babel-plugin-replace-textdomain 9. Missing context and domain parameters, no replacement: debug calls 1`] = `
[
  [
    Domain argument (index 3) is missing
> 1 | _x( 'No domain' );
    | ^^^^^^^^^^^^^^^^^,
  ],
]
`;

exports[`@automattic/babel-plugin-replace-textdomain 10. Non-literal domain: 10. Non-literal domain 1`] = `

__( 'Non-literal domain', domain );

      ↓ ↓ ↓ ↓ ↓ ↓

__('Non-literal domain', domain);

`;

exports[`@automattic/babel-plugin-replace-textdomain 10. Non-literal domain: debug calls 1`] = `
[
  [
    Domain argument should be a StringLiteral, not Identifier
> 1 | __( 'Non-literal domain', domain );
    |                           ^^^^^^,
  ],
]
`;

exports[`@automattic/babel-plugin-replace-textdomain 11. Template-string domain: 11. Template-string domain 1`] = `

__( 'Template-string domain', \`domain\` );

      ↓ ↓ ↓ ↓ ↓ ↓

__('Template-string domain', 'new-domain');

`;

exports[`@automattic/babel-plugin-replace-textdomain 11. Template-string domain: debug calls 1`] = `[]`;

exports[`@automattic/babel-plugin-replace-textdomain 12. Template-string domain, no replacement: 12. Template-string domain, no replacement 1`] = `

__( 'Template-string domain', \`domain\` );

      ↓ ↓ ↓ ↓ ↓ ↓

__('Template-string domain', \`domain\`);

`;

exports[`@automattic/babel-plugin-replace-textdomain 12. Template-string domain, no replacement: debug calls 1`] = `
[
  [
    No mapping for textdomain domain (first instance)
> 1 | __( 'Template-string domain', \`domain\` );
    |                               ^^^^^^^^,
  ],
]
`;

exports[`@automattic/babel-plugin-replace-textdomain 13. Template-string domain with expression: 13. Template-string domain with expression 1`] = `

__( 'Template-string domain', \`domain \${ x }\` );

      ↓ ↓ ↓ ↓ ↓ ↓

__('Template-string domain', \`domain \${x}\`);

`;

exports[`@automattic/babel-plugin-replace-textdomain 13. Template-string domain with expression: debug calls 1`] = `
[
  [
    Domain argument should be a StringLiteral, not TemplateLiteral
> 1 | __( 'Template-string domain', \`domain \${ x }\` );
    |                               ^^^^^^^^^^^^^^^,
  ],
]
`;

exports[`@automattic/babel-plugin-replace-textdomain 14. Expression domain: 14. Expression domain 1`] = `

__( 'Expression', 'dom' + 'ain' );

      ↓ ↓ ↓ ↓ ↓ ↓

__('Expression', 'dom' + 'ain');

`;

exports[`@automattic/babel-plugin-replace-textdomain 14. Expression domain: debug calls 1`] = `
[
  [
    Domain argument should be a StringLiteral, not BinaryExpression
> 1 | __( 'Expression', 'dom' + 'ain' );
    |                   ^^^^^^^^^^^^^,
  ],
]
`;

exports[`@automattic/babel-plugin-replace-textdomain 15. Doesn't try to handle \`toString()\` or the like: debug calls 1`] = `[]`;
